### DISCUSSION TITLE
Vimeo and Microsoft Stream filtering

### DISCUSSION THREAD

**Discussion by bee911** [FrostCo/AdvancedProfanityFilter#404] [General]
Thank you very much for this filter.  works perfectly with web sites and youtube, amazon, etc. video sites.  however, i am unable to get to work with vimeo videos, and don't see vimeo listed under config\audio\supported sites.  with closed caption on, advancedprofanityfilter will successfully substitute closed-captioned text found with word i configured (*REMOVED*) as shown in below image

![image](https://user-images.githubusercontent.com/62043136/165982028-0a7af1a9-4ad9-4013-a19a-17973a15b777.png)

but it will not mute audio as it does with other supported sites.  i have reviewed your doc (Audio Muting: Adding a Custom Site) but unable to find a word in the f12 elements second, and have also attempted 3. Video TextTrack Mode method without success.  i am using chrome browser

finally, i am also not able to get it to work with microsoft stream (web.microsoftstream.com) but did find the node as your doc stated as shown below:

![image](https://user-images.githubusercontent.com/62043136/165991562-0ccf59c8-6aa6-4988-8ad3-6b5d5bc9ad76.png)

would you please let me know how to add filtering for microsoft stream as well?  thank you much for help.

**Comment by richardfrost** [comment]
You're welcome @bee911! I'm so glad you like it.

# Vimeo
Here are the details I found for Vimeo. I tested with this [video](https://vimeo.com/703355279). Temporarily adding the word `summertime` to the filter, and testing around the 4:10 mark.

I looked through several videos, and I never found one that had multiple lines in the captions. If there is a video that has multiple lines, I'd like to see it, because I don't think this config will work for it, but it does seem to work fine as long as there is only one line. If you can, take it for a spin and let me know how it goes.

### HTML Structure
```html
<div class="vp-captions opacity-100 color-black" aria-live="assertive" style="font-size: 24px; font-variant: initial; font-family: &quot;Helvetica Neue&quot;, Helvetica, Arial, sans-serif; color: rgb(255, 255, 255); text-shadow: none;" lang="en-x-autogenerated" dir="ltr">
  <span class="vp-captions-window" style="background-color: rgba(0, 0, 0, 0);">
    <span class="vp-captions-line"> of this recliner summertime, you know, what's outside what sun</span> <!-- node -->
  </span>
</div>
```

### Audio Config
```json
{
  "vimeo.com": {
    "className": "vp-captions-line",
    "displaySelector": "div.vp-captions > span.vp-captions-window",
    "mode": "element",
    "note": "Only tested with single line of caption/subtitle text",
    "tagName": "SPAN"
  }
}
```

To test this, you can copy-paste this to the "Custom Sites" area found in the extensions Option's page > Audio tab. After pasting it in there, click the blue "Save" button, and then refresh/open a new tab to the video site and give it a try. If it works, I'll go ahead and add it to the official list of supported sites.

# Microsoft Stream
I don't have access to this service, but the details you included are definitely helpful. Unfortunately there isn't quite enough to on on with the screenshot you gave (which is definitely valuable, but doesn't have quite enough information). If you could get the element just above that one it would be helpful. You can probably just play a video with captions, and then use the web developer tools on the elements page, and search for the caption text and then send me a screenshot of the surrounding elements. Really I just need to know the parent, or maybe grandparent of the example node you posted. If you can get that to me I should be able to get a config for you to test.

**Comment by bee911** [comment]
hello, thank you so much for you reply.  i greatly appreciate your help.

unfortunately, the vimeo video is still not working consistently.  as appears my investment site has player in an iframe. here more details from f12 dev tools if helpful.  highlighted is foul language:

![image](https://user-images.githubusercontent.com/62043136/166124118-9cabe70c-c60e-461a-bb80-29335e0f80ee.png)

i was able to get vimeo to work once i saw the vimeo url above.  so i simply changed the config you provided prefixing player. and works perfectly:

```json
{
  "player.vimeo.com": {
    "className": "vp-captions-line",
    "displaySelector": "div.vp-captions > span.vp-captions-window",
    "mode": "element",
    "note": "Only tested with single line of caption/subtitle text",
    "tagName": "SPAN"
  }
}
```

here are details from microsoft stream (web.microsoftstream.com):
![image](https://user-images.githubusercontent.com/62043136/166124367-e35fb235-71f1-4c96-9daa-f5d5957ac364.png)

**Reply by bee911** [reply]
followup to reply i just made...it appears vimeo is filtering really bad language in the CC as shown below as that isn't your filter.  in your filter config i have it set to replace bad word with REMOVED which it always does.  but as shown below you see the badword is b*******.   words that are in your wordlist are replaced with REMOVED and your filter does beep it out.  :)  but since vimeo is filtering out these horrible words before your filter can see it, your filter cannot beep the sound.  :(

![image](https://user-images.githubusercontent.com/62043136/166124694-8b3c0476-eb4c-456f-9435-0ce81b0cfc79.png)

however, i was just able to get your filter to work with vimeo by adding * to wordlist:

![image](https://user-images.githubusercontent.com/62043136/166125124-b4fc08e9-475c-4af4-9ea5-b559363cc2a6.png)

once i did your filter beeped the sound perfectly and replaced CC with REMOVED as shown below!  Praise the Lord!
![image](https://user-images.githubusercontent.com/62043136/166125169-5d0e95cc-b65b-4365-bfa3-af56676d2984.png)

so vimeo works perfectly now!  perhaps when you release the config for it something can be done so people don't have to add * to wordlist since vimeo and microsoft stream are filtering out bad word in CC?

finally, microsoft stream is also doing same thing as vimeo for CC but is more restrictive as it doesn't even display first character but replaces bad word with all asterisks as shown below.

![image](https://user-images.githubusercontent.com/62043136/166124905-0af082dc-eb38-4709-b94a-c0d7923f8715.png)

thanks again and i look forward to hearing from you soon re config for microsoft stream.

**Reply by richardfrost** [reply]
Thanks for testing it, and finding all that info! And good job making that change to get it to work on the Vimeo video in the iframe.

Are you sure they are the ones making the changes to the captions? If you could try it on another browser or in Incognito mode without any extensions running just to confirm that would be helpful. Definitely not unheard of for services to do that (YouTube does it, for instance), but I just wanted to double check.

Here is a combined config that includes my best guess for Microsoft Stream with the new information:

```json
{
  "vimeo.com": [
    {
      "mode": "element",
      "tagName": "SPAN",
      "note": "Only works if captions are only on a single line",
      "className": "vp-captions-line",
      "displaySelector": "div.vp-captions > span.vp-captions-window"
    }
  ],
  "player.vimeo.com": [
    {
      "mode": "element",
      "tagName": "SPAN",
      "note": "Only works if captions are only on a single line",
      "className": "vp-captions-line",
      "displaySelector": "div.vp-captions > span.vp-captions-window"
    }
  ],
  "web.microsoftstream.com": [
    {
      "mode": "elementChild",
      "parentSelector": "div.video-player-container div.vjs-text-track-display",
      "subtitleSelector": "div > div",
      "tagName": "div"
    }
  ]
}
```
That one should work on either the iframe/embedded Vimeo video, one directly on their site, and the Microsoft Stream site.

**Reply by bee911** [reply]
indeed.  i tried in firefox and i had never access either site with it nor have i installed your filter in firefox.  that is how i came to conclusion as was shocked where i saw it working for some of my words but not ones it should have worked with.

i also just saw you had an iphone app which i just purchased.  thank you.  i used inline editor to copy my config and see all changes.  will it work with vimeo safari too as it does with chrome/edge on windows?

unfortunately, i was unable to get to work with microsoftstream with new config you provided.  it popped CC with REMOVED, but didn't beep sound.  :(

![image](https://user-images.githubusercontent.com/62043136/166126355-ce0b9767-3b98-4c08-8ea1-16c2333d369a.png)



thanks again for all your help and wonderful product!

**Reply by bee911** [reply]
hello richard, did you by chance see my reply a few days ago about microsoft stream config not working?

```json
{
  "web.microsoftstream.com": [
    {
      "mode": "elementChild",
      "parentSelector": "div.video-player-container div.vjs-text-track-display",
      "subtitleSelector": "div > div",
      "tagName": "div"
    }
  ]
}
```

profanityfilter will correctly substitute text in CC, but it will not beep out sound.  please let me know if there is another config can do.  thank you.

**Reply by richardfrost** [reply]
@bee911 Sorry, yes I did see your latest reply. I've just been really busy and haven't had a chance to look into it more yet.

The Safari (iOS) version should work much like the regular extension, but there are some sites that don't allow you to play video on them, so the audio muting may not work on as many sites there. You can always give it a try though and see if it works for you.

As for them doing a pre-filtered text for their captions, that's unfortunate, but like I said, other services do it as well. In the case of those sites, you could setup a separate wordlist just for them, or change the default audio wordlist to include something, I would probably do a partial match word of `**` or `***` to get match those captions.

As for the Microsoft Stream one, I just double-checked, and it does seem like that config should be right.... Could you try getting the `node` again on the site to see if it is the same? Some sites will vary how the insert the captions, and so while you may have found one way, there could be another that we need to account for.

**Reply by bee911** [reply]
sorry to bother then.

here are few more captures from stream that will hopefully help.  my search term was  "what sort" which was before bad word.   thank you.

![image](https://user-images.githubusercontent.com/62043136/166675890-d8044481-f907-4668-83fd-23756c93b45f.png)

![image](https://user-images.githubusercontent.com/62043136/166676444-8333fcb1-14c2-45de-9c9d-11fb5efea9bf.png)

![image](https://user-images.githubusercontent.com/62043136/166676796-ef80cf34-362e-4c1b-959f-2a7263b03d6c.png)

**Reply by richardfrost** [reply]
You're not bugging me @bee911, I was just apologizing for the delay.

Thanks for the additional details. It looks like that matches what we had before, so perhaps its just the `node` that changes? Anyway, if you want to get me another example of the `node` details by following the steps [here](https://github.com/richardfrost/AdvancedProfanityFilter/wiki/Audio-Muting:-Adding-a-Custom-Site#2-elementtext-mode) I might be able to help get the `elementChild` mode working, but if not we can probably fall back to the `watcher` method:

```json
{
  "web.microsoftstream.com": [
    {
      "displaySelector": "",
      "iframe": false,
      "mode": "watcher",
      "subtitleSelector": "div.video-player-container div.vjs-text-track-display > div > div > div"
    }
  ]
}
```

**Reply by bee911** [reply]
thank you.  this config worked perfectly with microsoft stream.  are you still needing me to try the elementchild method or are we good to leave as is?

as mentioned last week, for both vimeo and stream i also had to add asterisk to wordlist since both services are filtering really bad words in CC before it gets to me and replacing.

![image](https://user-images.githubusercontent.com/62043136/166730006-6b0c691c-177f-4d66-a2eb-36b0d7b511b6.png)


will you be pushing both vimeo and stream config's  on next update or should i just leave as additional config going forward?

finally, i attempted to play same vimeo videos on iphone and unfortuantely it never beeps sound like it does on PC browsers.   i haven't yet tried microsoft stream.

take care

**Reply by richardfrost** [reply]
Excellent! I'm glad that worked. No, I don't think we need to keep going with the `elementChild` mode if you're good with the watcher. Just so you know the difference, the `element/elementChild` mode is usually preferrable because it will be slightly faster at catching things when compared with `watcher`, but sometimes `watcher` is the only one that will work for a site.

I do plan on adding the Vimeo config directly to the next version of the extension, so you will be able to remove it from your custom sites config once that gets out there. I probably won't add Microsoft Stream there, just because it doesn't seem to be a very public offering, where Vimeo is widely used.

For supporting muting for these two sites that like to filter captions, you may want to do one of the following:
1. If you don't mind having `**` be filtered on all pages, I would just add the word: `**` with the partial match method.
2. If you'd rather that word (`**`) only be filtered during audio filtering:
    1. Add the word `**` with the partial match method
    2. Choose a wordlist number (or give it a name) to be your default text filter list (Options page > Lists tab)
    3. Choose another wordlist number (or give it a name) to be your default audio filter list
    4. Using the bulk editor (Options page > Words tab), assign all the words except (`**`) to the default text filter wordlist
    5. Also using the bulk editor, set all the words including `**` to the audio filter wordlist
    6. Now, if `**` appears in text it won't be filtered unless its related to audio muting
    7. If you'd like, you can even setup a wordlist specifically to be used for audio muting for each site in question, but that might be overkill.

Not sure why it wouldn't work on the iPhone, but it did work playing a direct video (linked in my first post) on an iPad. I'll try to test it on an iPhone soon. Does it work on the video I linked above? It may be related to being in an iframe.

**Reply by bee911** [reply]
thank you much for reply.  perfect!  i wasn't aware could have different word list for audio so that would be great as i am finding ** is removed everywhere.  i did as you suggested and renamed wordlist1 and 2 to text and audio and set accordingly and propagated all words in bulk editor:

![image](https://user-images.githubusercontent.com/62043136/166916873-4ab414eb-1253-4fd3-9352-b680fe96586e.png)

![image](https://user-images.githubusercontent.com/62043136/166916677-641564e5-ce59-4da1-a4dc-65ee171eed05.png)

b/c of config i did above is "Default Wordlist" (below) always mean "Text" now?
![image](https://user-images.githubusercontent.com/62043136/166917328-4a2b7d64-24d5-4009-8fe4-f4461a7b72ad.png)

finally, do you think it would be possible for it to work in iphone/ipad or not hopeful?

one question, i have "God" added to wordlist as shown below.  but for some reason it is also filtering "good".  ??
![image](https://user-images.githubusercontent.com/62043136/166964105-c949ce97-c91c-48a5-bf47-54020d8c53f8.png)


thank you again!

**Reply by richardfrost** [reply]
You're welcome! Nice work, that's exactly what the wordlists are for.

Yes, you are correct on your thinking with the default wordlist shown. Basically it means that on the current domain you are using the default wordlist for filtering text, which means that the current domain doesn't have an override. You can assign a specific wordlist for a specific domain if you'd like. But by default, every domain will use the default unless you specifically tell it to use a different one.

For "God", is that word for sure set to "Exact" match method? As long as it is, I don't think that should be happening. Do you have another word that might be matching it? If you can send a screenshot of the example, let me know and I'll take a look.

As for iPhone/iPad, I'll have to do a bit more digging, and unfortunately I don't have easy access to those devices so it may take me a bit longer to investigate. Did you have a chance to try the video from my first reply to this discussion on your iPhone/iPad after adding the config? You can just add a word to your filter temporarily to test with that video. I'm curious to know if it works for you, as it did work with the iPad I tested with. It is very possible there might be an issue with the fact that it is in an iframe. I'll have to see if I can find a video from Vimeo that is embedded in an iframe somewhere.

**Reply by bee911** [reply]
yes, it is specified as "God" in filter config as shown in my previous screen snapshot above and "Exact".

here is one of my sites showing issue where you will see removed which is the word Good
![image](https://user-images.githubusercontent.com/62043136/167010195-93dec7ee-c467-47f7-ab73-ebf1f1349971.png)
![image](https://user-images.githubusercontent.com/62043136/167010348-5335d14b-f462-4f99-b17f-f54fb79ef428.png)

only way i can get to stop is i have to add the domain to filter config and set to disabled.  so here is what it should be:
![image](https://user-images.githubusercontent.com/62043136/167010677-09e26b58-0298-40a2-9a7d-9ebeca1fdd63.png)


sorry i didn't have chance to try your video yet.

thank you for help.

**Reply by richardfrost** [reply]
Oh, sorry. You have it set up correctly, but need to uncheck "Repeated" for "God". What is happening is its including words that repeat characters. So, for instance, "god" would match "ggoodd" as well as "good", which you found. That option is there because some people will add emphasis to curse words by repeating characters, and its not very easy to add every variation, which is what this option tries to help with. A lot of words its okay to leave it on, but for this one you will need to uncheck it.

No worries on the Video stuff, I'm sorry I don't have a better/faster way to test them out, but I'll get to it when I can. It would help if you could test that one when you have time, but no rush.

**Reply by bee911** [reply]
oh.  ok.  i also have added "Jesus" and it too has repeated option set and it works good.  so repeated should be disabled for all words?

i will test the vimeon on iphone in little bit and let you know.

**Reply by richardfrost** [reply]
I would say for most words repeated is good to have enabled, as long as it works well. "god" is an example of a word that doesn't work well with it enabled, but any others that you aren't having problem I'd leave it on, because then it will match more variations of the word.

**Reply by bee911** [reply]
ok.  thanks.  i made change to "God" and sites work great now.  thanks!

i also just tested the vimeo video (https://vimeo.com/703355279) and added summertime word with current config i have on iphone and it too didn't work.

what i noticed is it was still displaying all CC even though i have filter config to only show CC when a word is filtered.   no beep occured. :(
